tests/qos/test_tunnel_qos_remap.py

```diff
@@ -0,0 +1,173 @@
+import logging
+import pytest
+import time
+from tests.common.helpers.assertions import pytest_require, pytest_assert
+from tests.common.dualtor.mux_simulator_control import toggle_all_simulator_ports_to_lower_tor # lgtm[py/unused-import]
+from tests.common.dualtor.dual_tor_utils import upper_tor_host, lower_tor_host, dualtor_info, get_t1_active_ptf_ports, is_tunnel_qos_remap_enabled
+from tests.common.fixtures.ptfhost_utils import change_mac_addresses, run_icmp_responder, run_garp_service # lgtm[py/unused-import]
+import ptf.packet as scapy
+from ptf.mask import Mask
+from ptf import testutils
+from ptf.testutils import simple_tcp_packet, simple_ipv4ip_packet
+
+pytestmark = [
+    pytest.mark.topology('t0')
+]
+
+logger = logging.getLogger(__name__)
+
+SERVER_IP = "192.168.0.2"
+DUMMY_IP = "1.1.1.1"
+DUMMY_MAC = "aa:aa:aa:aa:aa:aa"
+
+@pytest.fixture(scope='module', autouse=True)
+def check_running_condition(tbinfo, duthost):
+    """
+    The test can only be running on tunnel_qos_remap enabled dualtor testbed
+    """
+    # Check dualtor topo
+    pytest_require("dualtor" in tbinfo["topo"]["name"], "Only run on dualtor testbed.", True)
+    
+    # Check tunnel_qos_remap is enabled
+    pytest_require(is_tunnel_qos_remap_enabled(duthost), "Only run when tunnel_qos_remap is enabled", True)
+
+
+def _build_testing_packet(src_ip, dst_ip, active_tor_mac, standby_tor_mac, active_tor_ip, standby_tor_ip, inner_dscp, outer_dscp, ecn=1):
+    pkt = simple_tcp_packet(
+                eth_dst=standby_tor_mac,
+                ip_src=src_ip,
+                ip_dst=dst_ip,
+                ip_dscp=inner_dscp,
+                ip_ecn=ecn,
+                ip_ttl=64
+            )
+    # The ttl of inner_frame is decreased by 1
+    pkt.ttl -= 1
+    ipinip_packet = simple_ipv4ip_packet(
+                eth_dst=active_tor_mac,
+                eth_src=standby_tor_mac,
+                ip_src=standby_tor_ip,
+                ip_dst=active_tor_ip,
+                ip_dscp=outer_dscp,
+                ip_ecn=ecn,
+                inner_frame=pkt[IP]
+            )
+    pkt.ttl += 1
```

Comment:
Why is this needed?

Reply:
It's because the TTL of inner frame of encapsulated packet is decreased by 1. So we minus 1 at line 45. Then we need to recover it after building the ipinip packet.